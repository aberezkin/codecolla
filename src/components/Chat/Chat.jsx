import React, { Component } from 'react';
import Textarea from 'react-textarea-autosize';
import PropTypes from 'prop-types';
import './Chat.styl';
import Message from './Message/Message';

export const CHAT_CLASSNAME = 'Chat';

class Chat extends Component {
    constructor(props) {
        super(props);
        this.sendMessage = this.sendMessage.bind(this);
        this.onKeyDown = this.onKeyDown.bind(this);
        this.state = {
            textareaValue: '',
        };
    }

    onKeyDown(event) {
        if (event.keyCode === 13 && event.ctrlKey && this.state.textareaValue !== '')
            this.sendMessage();
    }

    sendMessage() {
        if (this.state.textareaValue === '') return;

        const elem = {
            author: this.props.nickname,
            content: this.state.textareaValue,
            date: new Date(),
        };

        this.setState({ textareaValue: '' });

        this.props.onMessage(elem);
    }

    render() {
        return (
            <div className={CHAT_CLASSNAME} style={this.props.style}>
                <div className="MessageBox">
                    {
                        this.props.messages.map(message => (
                            <Message
                                key={message.date.getTime()}
                                author={message.author}
                                content={message.content}
                                date={message.date}
                            />
                        ))
                    }
                </div>
                <div className="InputBox">
                    <Textarea
                        className="InputArea"
                        placeholder="Write your message..."
                        value={this.state.textareaValue}
                        onChange={event => this.setState({ textareaValue: event.target.value })}
                        onKeyDown={this.onKeyDown}
                        minRows={1}
                        maxRows={5}
                        tabIndex="0"
                    />
                    <div
                        className="SendButton"
                        onClick={this.sendMessage}
                        onKeyDown={this.onKeyDown}
                        role="button"
                        tabIndex="0"
                    >
                        <span>\&gt;</span>
                    </div>
                </div>
            </div>
        );
    }
}

Chat.propTypes = {
    style: PropTypes.objectOf(PropTypes.string),
    messages: PropTypes.arrayOf(PropTypes.object),
    onMessage: PropTypes.func,
    nickname: PropTypes.string.isRequired,
};

Chat.defaultProps = {
    style: {},
    messages: [],
    onMessage: () => { },
};

export default Chat;
