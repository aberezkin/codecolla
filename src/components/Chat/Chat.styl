.Chat
  min-width 320px
  height 100%
  display flex
  flex-direction column
  resize horizontal
  .MessageBox
    display block
    background-color lightcyan
    width 100%
    overflow-y auto
    flex 1 1 auto
  .InputBox
    height fit-content
    width 100%
    display flex
    flex-direction row
    .InputArea
      padding 4px
      width 100%
      resize none
      border none
      overflow auto
      outline none
      -webkit-box-shadow none
      -moz-box-shadow none
      box-shadow none
    .SendButton
      width 40px
      text-align center
      margin auto