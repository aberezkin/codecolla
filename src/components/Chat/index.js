import { connect } from 'react-redux';
import Chat from './Chat';
import { sendMessage } from '../../actions';

const mapStateToProps = state => ({
    messages: state.chat.messages,
    nickname: state.preferences.nickname,
});

const mapDispatchToProps = dispatch => ({
    onMessage: message => dispatch(sendMessage(message)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Chat);
