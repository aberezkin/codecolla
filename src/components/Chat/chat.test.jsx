import React from 'react';
import Chat from './Chat';

const setup = (propOverrides) => {
    const props = {
        messages: [],
        onMessage: jest.fn(),
        isVisible: true,
        nickname: 'john',
        ...propOverrides,
    };

    const wrapper = shallow((<Chat {...props} />));

    return {
        props,
        wrapper,
        send: wrapper.find('.SendButton'),
    };
};

describe('chat component', () => {
    it('should render chat', () => {
        const messages = [
            { author: 'me', content: 'Hello!', date: new Date('2017-12-22T14:53:13.141Z') },
            { author: 'you', content: 'Hi!', date: new Date('2017-12-22T14:53:13.241Z') },
            { author: 'me', content: 'How are you?', date: new Date('2017-12-22T14:53:14.141Z') },
        ];

        const { wrapper } = setup({ messages });

        expect(wrapper).toMatchSnapshot();
    });

    it('should send message when textarea value is empty on send button click', () => {
        const { props, send } = setup();
        send.simulate('click');
        expect(props.onMessage).not.toHaveBeenCalled();
    });

    it('should  when send button clicked', () => {
        const { props, send, wrapper } = setup();

        wrapper.setState({ textareaValue: 'some message' });
        send.simulate('click');
        expect(props.onMessage).toHaveBeenCalled();
    });

    it('should not send message when textarea value is empty on ctrl+enter'); // TODO: implement
    it('should send message when textarea value is not empty on ctrl+enter'); // TODO: implement
});
