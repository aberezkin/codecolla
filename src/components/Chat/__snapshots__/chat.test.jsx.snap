// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat component should render chat 1`] = `
<div
  className="Chat"
  style={Object {}}
>
  <div
    className="MessageBox"
  >
    <Message
      author="me"
      color="black"
      content="Hello!"
      date={2017-12-22T14:53:13.141Z}
      key="1513954393141"
    />
    <Message
      author="you"
      color="black"
      content="Hi!"
      date={2017-12-22T14:53:13.241Z}
      key="1513954393241"
    />
    <Message
      author="me"
      color="black"
      content="How are you?"
      date={2017-12-22T14:53:14.141Z}
      key="1513954394141"
    />
  </div>
  <div
    className="InputBox"
  >
    <TextareaAutosize
      className="InputArea"
      maxRows={5}
      minRows={1}
      onChange={[Function]}
      onHeightChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Write your message..."
      tabIndex="0"
      useCacheForDOMMeasurements={false}
      value=""
    />
    <div
      className="SendButton"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex="0"
    >
      <span>
        \\&gt;
      </span>
    </div>
  </div>
</div>
`;
